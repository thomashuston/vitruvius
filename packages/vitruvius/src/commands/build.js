import * as path from 'path';
import buildPackage from 'vitruvius-build-package';
import { getPackages } from 'vitruvius-lerna';

export const command = 'build';

export const describe = 'build packages';

export const builder = {
    src: {
        demandOption: true,
        describe: 'the source directory name',
        type: 'string'
    },
    dest: {
        demandOption: true,
        describe: 'the output directory name',
        type: 'string'
    },
    ignore: {
        default: [],
        demandOption: false,
        describe: 'glob patterns of files to ignore',
        type: 'array'
    }
};

export const handler = (argv) => {
    const packages = getPackages();
    packages.forEach((pkg) => {
        const srcDir = path.join(pkg, argv.src);
        const destDir = path.join(pkg, argv.dest);
        buildPackage({
            srcDir,
            destDir,
            ignorePatterns: argv.ignore
        });
    });
};
