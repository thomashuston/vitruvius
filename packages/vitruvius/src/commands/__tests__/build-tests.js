import * as path from 'path';
import { transform } from 'babel-core';
import buildPackage from 'vitruvius-build-package';
import { handler as build } from '../build';

let mockPackages;
jest.mock('vitruvius-lerna', () => ({
    getPackages: () => mockPackages
}));

jest.mock('vitruvius-build-package', () => jest.fn());

let output;
process.stdout.write = (text) => { output.push(text); };
process.exit = jest.fn();

const now = Date.now();
const rootDir = path.resolve(path.join(__dirname, '..', '..', '..', '..', '..'));

beforeEach(() => {
    output = [];
    buildPackage.mockReset();
    Date.now = () => now;
});

it('builds code in the source directory into the destination directory for all packages', () => {
    mockPackages = [
        path.join(rootDir, '/packages/fake-pkg-1'),
        path.join(rootDir, '/packages/fake-pkg-2')
    ];

    build({
        src: 'src',
        dest: 'lib',
        'ignore-files': [
            '**/__mocks__/**',
            '**/__tests__/**'
        ]
    });

    mockPackages.forEach((pkg) => {
        expect(buildPackage).toHaveBeenCalledWith({
            srcDir: path.join(pkg, 'src'),
            destDir: path.join(pkg, 'lib'),
            ignorePatterns: [
                '**/__mocks__/**',
                '**/__tests__/**'
            ]
        });
    });
});

it('does not build code for ignore packages', () => {
    mockPackages = [
        path.join(rootDir, '/examples/example-pkg'),
        path.join(rootDir, '/packages/fake-pkg-1'),
        path.join(rootDir, '/packages/fake-pkg-2')
    ];

    build({
        src: 'src',
        dest: 'lib',
        'ignore-packages': [
            'examples/**'
        ]
    });

    expect(buildPackage).toHaveBeenCalledTimes(2);
    expect(buildPackage).toHaveBeenCalledWith({
        srcDir: path.join(mockPackages[1], 'src'),
        destDir: path.join(mockPackages[1], 'lib')
    });
    expect(buildPackage).toHaveBeenCalledWith({
        srcDir: path.join(mockPackages[2], 'src'),
        destDir: path.join(mockPackages[2], 'lib')
    });
});

it('exits with success code when all packages build successfully', () => {
    mockPackages = [
        path.join(rootDir, '/packages/fake-pkg-1'),
        path.join(rootDir, '/packages/fake-pkg-2')
    ];

    build({
        src: 'src',
        dest: 'lib'
    });

    expect(process.exit).toHaveBeenCalledWith(0);
});

it('exits with failure code when some packages fail to build', () => {
    mockPackages = [
        path.join(rootDir, '/packages/fake-pkg-1'),
        path.join(rootDir, '/packages/fake-pkg-2')
    ];

    buildPackage.mockImplementationOnce(() => { throw new Error('failed!'); });

    build({
        src: 'src',
        dest: 'lib'
    });

    expect(process.exit).toHaveBeenCalledWith(1);
});

if (process.platform !== 'win32') {

    it('logs success when all packages build successfully', () => {
        mockPackages = [
            path.join(rootDir, '/packages/fake-pkg-1'),
            path.join(rootDir, '/packages/fake-pkg-2')
        ];

        build({
            src: 'src',
            dest: 'lib'
        });

        expect(output.join('')).toMatchSnapshot();
    });

    it('logs errors when some packages fail to build', () => {
        mockPackages = [
            path.join(rootDir, '/packages/fake-pkg-1'),
            path.join(rootDir, '/packages/fake-pkg-2')
        ];

        buildPackage.mockImplementationOnce(() => { throw new Error('failed!'); });

        build({
            src: 'src',
            dest: 'lib'
        });

        expect(output.join('')).toMatchSnapshot();
    });

    it('pretty prints babel errors when babel compilation fails in a package', () => {
        mockPackages = [
            path.join(rootDir, '/packages/fake-pkg-1'),
            path.join(rootDir, '/packages/fake-pkg-2')
        ];

        buildPackage.mockImplementationOnce(() => {
            transform('cons invalid = 1;', {
                filename: path.join(rootDir, '/packages/fake-pkg-1/invalid.js')
            });
        });

        build({
            src: 'src',
            dest: 'lib'
        });

        expect(output.join('')).toMatchSnapshot();
    });

}
