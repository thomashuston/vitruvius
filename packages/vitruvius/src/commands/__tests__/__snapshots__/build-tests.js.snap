// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logs errors when some packages fail to build 1`] = `
"[999D[K[0m[7m[33m[1m BUILD [22m[39m[27m[0m packages/fake-pkg-1[999D[K[0m[7m[31m[1m ERROR [22m[39m[27m[0m packages/fake-pkg-1

  Error: failed!

[999D[K[0m[7m[33m[1m BUILD [22m[39m[27m[0m packages/fake-pkg-2[999D[K[0m[7m[32m[1m BUILT [22m[39m[27m[0m packages/fake-pkg-2

[1mPackages: [22m[0m[31m[1m1 failed, [22m[39m[0m[0m[32m[1m1 built, [22m[39m[0m2 total
[1mTime:     [22m[0m[33m[1m0.00s
[22m[39m[0m[0m[31m[1mFailed to build 1 package.
[22m[39m[0m"
`;

exports[`logs success when all packages build successfully 1`] = `
"[999D[K[0m[7m[33m[1m BUILD [22m[39m[27m[0m packages/fake-pkg-1[999D[K[0m[7m[32m[1m BUILT [22m[39m[27m[0m packages/fake-pkg-1
[999D[K[0m[7m[33m[1m BUILD [22m[39m[27m[0m packages/fake-pkg-2[999D[K[0m[7m[32m[1m BUILT [22m[39m[27m[0m packages/fake-pkg-2

[1mPackages: [22m[0m[32m[1m2 built, [22m[39m[0m2 total
[1mTime:     [22m[0m[33m[1m0.00s
[22m[39m[0m[2mBuilt all packages.
[22m"
`;

exports[`pretty prints babel errors when babel compilation fails in a package 1`] = `
"[999D[K[0m[7m[33m[1m BUILD [22m[39m[27m[0m packages/fake-pkg-1[999D[K[0m[7m[31m[1m ERROR [22m[39m[27m[0m packages/fake-pkg-1

  Unexpected token, expected ;

   at [0m[36mpackages/fake-pkg-1/invalid.js[39m[0m:1:5

  [0m[31m[1m>[22m[39m[90m 1 | [39mcons invalid [33m=[39m [35m1[39m[33m;[39m
   [90m   | [39m     [31m[1m^[22m[39m[0m

[999D[K[0m[7m[33m[1m BUILD [22m[39m[27m[0m packages/fake-pkg-2[999D[K[0m[7m[32m[1m BUILT [22m[39m[27m[0m packages/fake-pkg-2

[1mPackages: [22m[0m[31m[1m1 failed, [22m[39m[0m[0m[32m[1m1 built, [22m[39m[0m2 total
[1mTime:     [22m[0m[33m[1m0.00s
[22m[39m[0m[0m[31m[1mFailed to build 1 package.
[22m[39m[0m"
`;
